discord: Various stack traces

a) discord app isn't running/crashes w/clientid:

```
 2023-01-13T15:54:00-0800 72470 discordbot/DiscordBot discordbot:_setup_ipc_client:66 ERROR Cannot configure IPC client: [Errno 61] Connection refused Traceback (most recent call last):
  File "/Users/aw/Src/whats-now-playing/nowplaying/processes/discordbot.py", line 64, in _setup_ipc_client
    await self.client['ipc'].connect()
  File "/Users/aw/wnpvenv/lib/python3.10/site-packages/pypresence/presence.py", line 80, in connect
    await self.handshake()
  File "/Users/aw/wnpvenv/lib/python3.10/site-packages/pypresence/baseclient.py", line 107, in handshake
    self.sock_reader, self.sock_writer = await asyncio.open_unix_connection(self.ipc_path)
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/streams.py", line 98, in open_unix_connection
    transport, _ = await loop.create_unix_connection(
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/unix_events.py", line 255, in create_unix_connection
    await self.sock_connect(sock, path)
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/selector_events.py", line 500, in sock_connect
    return await fut
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/selector_events.py", line 505, in _sock_connect
    sock.connect(address)
ConnectionRefusedError: [Errno 61] Connection refused
```

b) discord issue w/token

```
2023-01-13T15:54:03-0800 72470 discordbot/DiscordBot discordbot:update_activity:127 DEBUG Traceback (most recent call last):
  File "/Users/aw/Src/whats-now-playing/nowplaying/processes/discordbot.py", line 122, in update_activity
    await func(templateout)
  File "/Users/aw/Src/whats-now-playing/nowplaying/processes/discordbot.py", line 84, in _update_ipc
    await self.client['ipc'].update(state='Streaming', details=templateout)
  File "/Users/aw/wnpvenv/lib/python3.10/site-packages/pypresence/presence.py", line 70, in update
    self.send_data(1, payload)
  File "/Users/aw/wnpvenv/lib/python3.10/site-packages/pypresence/baseclient.py", line 96, in send_data
    assert self.sock_writer is not None, "You must connect your client before sending events!"
AssertionError: You must connect your client before sending events!
```

--- a/nowplaying/processes/discordbot.py
+++ b/nowplaying/processes/discordbot.py
@@ -39,6 +39,10 @@ class DiscordSupport:
         token = self.config.cparser.value('discord/token')
         if not token:
             return
+
+        if self.client.get('bot'):
+            return
+
         try:
             intents = discord.Intents.default()
             self.client['bot'] = discord.Client(intents=intents)
@@ -51,6 +55,9 @@ class DiscordSupport:
         task = loop.create_task(self.client['bot'].connect(reconnect=True))
         self.tasks.add(task)
         task.add_done_callback(self.tasks.discard)
+        while not self.stopevent.is_set() and not self.client['bot'].is_ready(
+        ):
+            await asyncio.sleep(1)
         logging.debug('bot setup')
 
     async def _setup_ipc_client(self):
@@ -62,9 +69,14 @@ class DiscordSupport:
         try:
             self.client['ipc'] = pypresence.AioPresence(clientid, loop=loop)
             await self.client['ipc'].connect()
+        except ConnectionRefusedError:
+            logging.error('Cannot connect to discord client.')
+            del self.client['ipc']
+            return
         except Exception as error:  #pylint: disable=broad-except
             logging.error('Cannot configure IPC client: %s %s', error,
                           traceback.format_exc())
+            del self.client['ipc']
             return
         logging.debug('ipc setup')
 
@@ -78,79 +90,75 @@ class DiscordSupport:
                 url=f'https://twitch.tv/{channelname}')
         else:
             activity = discord.Game(templateout)
-        await self.client['bot'].change_presence(activity=activity)
+        try:
+            await self.client['bot'].change_presence(activity=activity)
+        except ConnectionResetError:
+            logging.error('Cannot connect to discord.')
+            del self.client['bot']
 
     async def _update_ipc(self, templateout):
-        await self.client['ipc'].update(state='Streaming', details=templateout)
+        try:
+            await self.client['ipc'].update(state='Streaming',
+                                            details=templateout)
+        except ConnectionRefusedError:
+            logging.error('Cannot connect to discord client.')
+            del self.client['ipc']
+        except Exception as error:  #pylint: disable=broad-except
+            logging.error('Cannot configure IPC client: %s %s', error,
+                          traceback.format_exc())
+            del self.client['ipc']
 
-    async def update_activity(self):
-        ''' keep the bot's presence updated '''
+    async def connect_clients(self):
+        ''' (re-)connect clients '''
+        client = {
+            'bot': self._setup_bot_client,
+            'ipc': self._setup_ipc_client,
+        }
+
+        for mode, func in client.items():  # pylint: disable=consider-using-dict-items
+            if not self.client.get(mode):
+                await func()
+
+    async def start(self):
+        ''' start the service '''
 
         client = {
             'bot': self._update_bot,
             'ipc': self._update_ipc,
         }
 
-        try:
-            metadb = nowplaying.db.MetadataDB()
-            watcher = metadb.watcher()
-            watcher.start()
+        metadb = nowplaying.db.MetadataDB()
+        watcher = metadb.watcher()
+        watcher.start()
 
-            mytime = 0
-            while not self.stopevent.is_set():
-                await asyncio.sleep(1)
+        mytime = 0
 
-                if self.client.get(
-                        'bot') and not self.client['bot'].is_ready():
+        while not self.stopevent.is_set():
+            await self.connect_clients()
+            # discord will lock out if updates more than every 15 seconds
+            await asyncio.sleep(15)
+
+            if mytime < watcher.updatetime:
+                template = self.config.cparser.value('discord/template')
+                if not template:
                     continue
 
-                if mytime < watcher.updatetime:
-                    template = self.config.cparser.value('discord/template')
-                    if not template:
-                        continue
+                metadata = metadb.read_last_meta()
+                if not metadata:
+                    continue
 
-                    metadata = metadb.read_last_meta()
-                    if not metadata:
-                        continue
-
-                    templatehandler = nowplaying.utils.TemplateHandler(
-                        filename=template)
-                    mytime = watcher.updatetime
-                    templateout = templatehandler.generate(metadata)
-                    for mode, func in client.items():
-                        if self.client.get(mode):
+                templatehandler = nowplaying.utils.TemplateHandler(
+                    filename=template)
+                mytime = watcher.updatetime
+                templateout = templatehandler.generate(metadata)
+                for mode, func in client.items():
+                    if self.client.get(mode):
+                        try:
                             await func(templateout)
-                    # discord will lock out if updates more than every 15 seconds
-                    await asyncio.sleep(14)
 
-        except:  #pylint: disable=bare-except
-            logging.debug(traceback.format_exc())
-
-    async def start(self):
-        ''' start the service '''
-
-        client = {
-            'bot': self._setup_bot_client,
-            'ipc': self._setup_ipc_client,
-        }
-
-        while not self.stopevent.is_set():
-            await asyncio.sleep(10)
-            for func in client.values():  # pylint: disable=consider-using-dict-items
-                await func()
-            if self.client:
-                break
-
-        if self.stopevent.is_set():
-            return
-
-        loop = asyncio.get_running_loop()
-        task = loop.create_task(self.update_activity())
-        self.tasks.add(task)
-        task.add_done_callback(self.tasks.discard)
-
-        while not self.stopevent.is_set():
-            await asyncio.sleep(1)
+                        except:  #pylint: disable=bare-except
+                            logging.debug(traceback.format_exc())
+                            del self.client[mode]
         if self.client.get('bot'):  # pylint: disable=consider-using-dict-items
             await self.client['bot'].close()
 

